### ISSUE TITLE
Validator: No longer warns about CSP

### ISSUE THREAD

**Issue by erosman** [mozilla/addons-linter#1506]
Since the latest push, the validator no longer warns about CSP.

Example:
`https://addons.mozilla.org/editors/review/850282/`

**Comment by EnTeQuAk** [comment]
As far as I know we explicitly enabled them again since that matches current add-on policies.

See https://github.com/mozilla/addons-linter/blob/master/tests/parsers/test.manifestjson.js#L423-L426

Also… to verify issues faster and give more proper responses it'd be wonderful if you could add a little bit more detail - e.g what rules weren't warned about but you expected them to etc.

**Comment by erosman** [comment]
The issue should have been apparent from the link I gave. 

Example addon:
`https://addons.mozilla.org/editors/review/850282/`

Validator: No mention of CSP
`https://addons.mozilla.org/en-US/developers/addon/talking-webb/file/714747/validation`

manifest.json
`https://addons.mozilla.org/en-US/firefox/files/browse/714747/file/manifest.json#top`

Issue
`"content_security_policy": "script-src 'self' 'unsafe-eval'; object-src 'self'",`

Problem:
Validator No longer warns about existence of `CSP` and `unsafe-eval` 

I have my own scripts that picks up `CSP` and `unsafe-eval` but others may miss it. As the example clearly shows, the validator did not pick up the positional security risk of `CSP` & `unsafe-eval`

> Closing as this is expected and not a bug.
I didn't understand. Is it expected to miss unsafe-eval in CSP in validator?

**Comment by EnTeQuAk** [comment]
> Is it expected to miss unsafe-eval in CSP in validator?

As far as I know it's more that there are valid cases where it might make sense. Summoning @wagnerand

**Comment by wagnerand** [comment]
I just re-read the pull request discussions. The reason, I asked for this (in https://github.com/mozilla/addons-linter/pull/1434#pullrequestreview-54312954) was that `unsafe-eval` is not forbidden, which is what the rest of those tests is about. 

That said, @erosman is probably right in that we might want to add a warning for `unsafe-eval` (note that we can ignore `unsafe-inline`, since it's unsupported as I learned in the meantime). But that warning should be about `eval` and is unrelated to remote script execution, which is what that patch was about (https://github.com/mozilla/addons-linter/pull/1434/files#diff-abf32190003257adb7218839b0bce1ffR52).

**Comment by kewisch** [comment]
Can you share more details on why unsafe-inline is unimportant? AIUI, if there is code that manages to inject DOM elements from a remote source, this could also be a script tag, which unsafe-inline will prevent. Sure, this may not be as common, but as the workaround is just making that script a separate file I don't see why we should relax this rule.

**Comment by wagnerand** [comment]
https://developer.mozilla.org/en-US/Add-ons/WebExtensions/manifest.json/content_security_policy says that `unsafe-inline` is entirely unsupported.

**Comment by EnTeQuAk** [comment]
@wagnerand we don't have any way of saying that `unsafe-inline` is unsupported, so maybe flagging it might be worth doing too now that we aim to flag `unsafe-eval` too? Especially once it get's supported?

**Comment by wagnerand** [comment]
@EnTeQuAk I am inclined to say we just ignore `unsafe-inline`.

Also, let me find out whether we plan on supporting it at some point.

**Comment by wagnerand** [comment]
According to @kmaglione, we don't plan on supporting it.

**Comment by erosman** [comment]
Another example
```
https://addons.mozilla.org/editors/review/851577/
https://addons.mozilla.org/en-US/firefox/files/browse/717300/file/manifest.json#top
"content_security_policy": "script-src 'self' 'unsafe-eval'; object-src 'self'"
```

**Comment by ValentinaPC** [comment]
Verified this on AMO-dev FF55(Win 7).
If I understood well we should show a warning for unsefe-eval as in the example provided above, but we don't. I misunderstood?
Attaching used webextension.
[eval in CSP.zip](https://github.com/mozilla/addons-linter/files/1287913/eval.in.CSP.zip)

**Comment by erosman** [comment]
> If I understood well we should show a warning for unsefe-eval as in the example provided above, but we don't. I misunderstood?
There used to be a warning about CSP, regardless of its content.
For example here:
`https://addons.mozilla.org/en-US/developers/addon/preview-app-v2/file/704122/validation`

Of course, that warning was not always needed since CSP in itself is not a problem.

From what I understand, the validator was updated to only show a warning when CSP exists AND it contains `unsefe-eval`, `URL`, etc.

The updated validator now misses the CSP altogether as it the examples given.

**Comment by ValentinaPC** [comment]
Just to be sure....

CSP -  no unsave eval => warning
![csp_-_no_eval](https://user-images.githubusercontent.com/15685960/30213421-a1af10ce-94b1-11e7-8954-2dd7398fc220.png)

CSP - with unsave eval => no warning
![csp-eval](https://user-images.githubusercontent.com/15685960/30213441-aed553da-94b1-11e7-85e2-c55faf73dac4.png)

Are ok?

**Comment by wagnerand** [comment]
Doesn't seem right to me.

The linter should raise a warning (respectively) for
1) Using a CSP that allows remote scripts, for example: `script-src https:;`
2) Using a CSP that allows eval, for example: `script-src 'self' 'unsafe-eval;`

So, the CSP `script-src https: 'unsafe-eval'` should raise two warnings.

**Comment by wagnerand** [comment]
Also, this works for me locally using the file from https://github.com/mozilla/addons-linter/issues/1506#issuecomment-328097502.

```
 ./bin/addons-linter ~/Downloads/eval.in.CSP.zip
Validation Summary:

errors          0
notices         0
warnings        1

WARNINGS:

Code                       Message                                            Description                                                                                       File            Line   Column
MANIFEST_CSP_UNSAFE_EVAL   Using 'eval' has strong security and performance   In most cases the same result can be achieved differently, therefore it is generally prohibited   manifest.json
                           implications.
```

@ValentinaPC Did you test on -dev? This isn't on -dev yet.

**Comment by EnTeQuAk** [comment]
I'll update the issue once this is on dev, I'll probably release a new version Monday morning.

**Comment by ValentinaPC** [comment]
@EnTeQuAk : not yet on -dev? Thanks!

**Comment by EnTeQuAk** [comment]
Ah, I forgot this issue. It is now, got a release yesterday. It's on -dev and -stage.

**Comment by ValentinaPC** [comment]
> So, the CSP script-src https: 'unsafe-eval'; should raise two warnings.

I cannot see 2 warnings. I see no warning at validation and a warning in the report page.
![1](https://user-images.githubusercontent.com/15685960/30375868-ce67dae4-9892-11e7-8935-c6f2da71cf89.png)
![report](https://user-images.githubusercontent.com/15685960/30375873-d18f60b6-9892-11e7-9896-7ee1cd5da1e9.png)

Used the same extension as before.

**Comment by wagnerand** [comment]
Yeah, I can confirm that. @EnTeQuAk do you have time to look into this?

**Comment by ValentinaPC** [comment]
Also, when I use both options in the form:
`"content_security_policy": "script-src 'self'  'unsafe-eval' https://example.com; object-src 'self'"`
I receive 1 warning at validation and 2 warnings in the report page.
If I reverse the order and use the form:
`"content_security_policy": "script-src 'self'  https://example.com; 'unsafe-eval' object-src 'self'"`
I receive 1 warning validation and 1 warning in the report page

**Comment by EnTeQuAk** [comment]
Mhmm, interesting. I'll look into it later today 


On Wed, Sep 13, 2017, at 02:08 PM, ValentinaPC wrote:
> Also, when I use both options in the form:
> "content_security_policy": "script-src 'self' 'unsafe-eval'
> https://example.com; object-src 'self'">  I receive 1 warning at validation and 2 warnings in the report page.>  If I reverse the order and use the form:
> "content_security_policy": "script-src 'self' https://example.com; 'unsafe-
> eval' object-src 'self'">  I receive 1 warning validation and 1 warning in the report page


> — You are receiving this because you were mentioned. Reply to this
> email directly, view it on GitHub[1], or mute the thread[2].> 

Links:

  1. https://github.com/mozilla/addons-linter/issues/1506#issuecomment-329147577

**Comment by EnTeQuAk** [comment]
@ValentinaPC @wagnerand yeah, I can confirm that and it looks similar (or even a duplicate) to https://github.com/mozilla/addons-linter/issues/1547

I'll use #1547 to track this and work on it there.
